File: cpp/libcudf_streaming/tests/streaming/test_approx_distinct_count.cpp
/**
 * SPDX-FileCopyrightText: Copyright (c) 2026, NVIDIA CORPORATION & AFFILIATES. All rights reserved.
 * SPDX-License-Identifier: Apache-2.0
 */

#include "../utils.hpp"
#include "base_streaming_fixture.hpp"

#include <cudf_test/column_wrapper.hpp>

#include <cudf/table/table.hpp>

#include <cudf_streaming/approx_distinct_count.hpp>
#include <cudf_streaming/table_chunk.hpp>

#include <gtest/gtest.h>
#include <rapidsmpf/error.hpp>
#include <rapidsmpf/streaming/core/actor.hpp>
#include <rapidsmpf/streaming/core/leaf_actor.hpp>

#include <cstdint>
#include <memory>
#include <vector>

namespace {

using CardinalityEstimatorTest = BaseStreamingFixture;

TEST_F(CardinalityEstimatorTest, EstimatesGlobalUnionAndCountsRows)
{
  constexpr std::size_t distinct_values = 10'000;
  constexpr std::size_t repetitions     = 2;
  auto values                           = iota_vector<std::int64_t>(distinct_values);
  auto const dups                       = values;
  values.insert(values.end(), dups.begin(), dups.end());

  cudf::test::fixed_width_column_wrapper<std::int64_t> column(values.begin(), values.end());
  std::vector<std::unique_ptr<cudf::column>> columns;
  columns.push_back(column.release());
  auto table = std::make_unique<cudf::table>(std::move(columns));

  std::vector<rapidsmpf::streaming::Message> inputs;
  inputs.push_back(cudf_streaming::to_message(
    0, std::make_unique<cudf_streaming::table_chunk>(std::move(table), stream)));

  auto ch_in  = ctx->create_channel();
  auto ch_out = ctx->create_channel();
  cudf_streaming::cardinality_estimator estimator{
    ctx, GlobalEnvironment->comm_, rapidsmpf::OpID{0}, 14};

  std::vector<rapidsmpf::streaming::Message> outputs;
  std::vector<rapidsmpf::streaming::Actor> actors;
  actors.push_back(rapidsmpf::streaming::actor::push_to_channel(ctx, ch_in, std::move(inputs)));
  actors.push_back(estimator.estimate(ch_in, ch_out));
  actors.push_back(rapidsmpf::streaming::actor::pull_from_channel(ctx, ch_out, outputs));

  rapidsmpf::streaming::run_actor_network(std::move(actors));

  ASSERT_EQ(outputs.size(), 1);
  auto const estimate = outputs.front().release<cudf_streaming::cardinality_estimate>();
  EXPECT_EQ(estimate.row_count,
            distinct_values * repetitions *
              rapidsmpf::safe_cast<std::size_t>(GlobalEnvironment->comm_->nranks()));
  EXPECT_NEAR(
    estimate.distinct_count, distinct_values, static_cast<double>(distinct_values) * 0.05);
}

TEST_F(CardinalityEstimatorTest, EmptyInput)
{
  auto ch_in  = ctx->create_channel();
  auto ch_out = ctx->create_channel();
  cudf_streaming::cardinality_estimator estimator{
    ctx, GlobalEnvironment->comm_, rapidsmpf::OpID{0}};

  std::vector<rapidsmpf::streaming::Message> outputs;
  std::vector<rapidsmpf::streaming::Actor> actors;
  actors.push_back(rapidsmpf::streaming::actor::push_to_channel(ctx, ch_in, {}));
  actors.push_back(estimator.estimate(ch_in, ch_out));
  actors.push_back(rapidsmpf::streaming::actor::pull_from_channel(ctx, ch_out, outputs));

  rapidsmpf::streaming::run_actor_network(std::move(actors));

  ASSERT_EQ(outputs.size(), 1);
  auto const estimate = outputs.front().release<cudf_streaming::cardinality_estimate>();
  EXPECT_EQ(estimate.row_count, 0);
  EXPECT_EQ(estimate.distinct_count, 0);
}

TEST_F(CardinalityEstimatorTest, SamplesSelectedColumnsAndForwardsInput)
{
  cudf::test::fixed_width_column_wrapper<std::int64_t> keys{0, 0, 1, 1};
  cudf::test::fixed_width_column_wrapper<std::int64_t> values{0, 1, 2, 3};
  std::vector<std::unique_ptr<cudf::column>> columns;
  columns.push_back(keys.release());
  columns.push_back(values.release());
  auto table = std::make_unique<cudf::table>(std::move(columns));

  std::vector<rapidsmpf::streaming::Message> inputs;
  inputs.push_back(cudf_streaming::to_message(
    7, std::make_unique<cudf_streaming::table_chunk>(std::move(table), stream)));

  auto ch_in      = ctx->create_channel();
  auto ch_sampled = ctx->create_channel();
  auto ch_out     = ctx->create_channel();
  cudf_streaming::cardinality_estimator estimator{
    ctx, GlobalEnvironment->comm_, rapidsmpf::OpID{0}, 14};

  std::vector<rapidsmpf::streaming::Message> sampled;
  std::vector<rapidsmpf::streaming::Message> estimates;
  std::vector<rapidsmpf::streaming::Actor> actors;
  actors.push_back(rapidsmpf::streaming::actor::push_to_channel(ctx, ch_in, std::move(inputs)));
  actors.push_back(estimator.estimate(ch_in, ch_out, ch_sampled, {0}));
  actors.push_back(rapidsmpf::streaming::actor::pull_from_channel(ctx, ch_sampled, sampled));
  actors.push_back(rapidsmpf::streaming::actor::pull_from_channel(ctx, ch_out, estimates));

  rapidsmpf::streaming::run_actor_network(std::move(actors));

  ASSERT_EQ(sampled.size(), 1);
  EXPECT_EQ(sampled.front().sequence_number(), 7);
  ASSERT_EQ(estimates.size(), 1);
  auto const estimate = estimates.front().release<cudf_streaming::cardinality_estimate>();
  EXPECT_EQ(estimate.row_count,
            4 * rapidsmpf::safe_cast<std::size_t>(GlobalEnvironment->comm_->nranks()));
  EXPECT_NEAR(estimate.distinct_count, 2, 1);
}

TEST_F(CardinalityEstimatorTest, RejectsInvalidPrecision)
{
  EXPECT_THROW(
    cudf_streaming::cardinality_estimator(ctx, GlobalEnvironment->comm_, rapidsmpf::OpID{0}, 3),
    std::invalid_argument);
  EXPECT_THROW(
    cudf_streaming::cardinality_estimator(ctx, GlobalEnvironment->comm_, rapidsmpf::OpID{0}, 19),
    std::invalid_argument);
}

}  // namespace

File: python/cudf_streaming/cudf_streaming/tests/test_approx_distinct_count.py
# SPDX-FileCopyrightText: Copyright (c) 2026, NVIDIA CORPORATION & AFFILIATES. All rights reserved.
# SPDX-License-Identifier: Apache-2.0

from __future__ import annotations

from typing import TYPE_CHECKING

import numpy as np
import pylibcudf as plc
import pytest

from cudf_streaming import CardinalityEstimate, CardinalityEstimator
from cudf_streaming.table_chunk import TableChunk
from rapidsmpf.streaming.core.actor import run_actor_network
from rapidsmpf.streaming.core.leaf_actor import (
    pull_from_channel,
    push_to_channel,
)
from rapidsmpf.streaming.core.message import Message

if TYPE_CHECKING:
    from rapidsmpf.communicator.communicator import Communicator
    from rapidsmpf.streaming.core.context import Context


def make_table(values: np.ndarray, context: Context) -> TableChunk:
    stream = context.br().stream_pool.get_stream()
    table = plc.Table([plc.Column.from_array(values, stream=stream)])
    return TableChunk.from_pylibcudf_table(
        table, stream, exclusive_view=True, br=context.br()
    )


def estimate(
    context: Context,
    comm: Communicator,
    chunks: list[TableChunk],
    *,
    precision: int = 14,
) -> CardinalityEstimate:
    estimator = CardinalityEstimator(context, comm, tag=0, precision=precision)
    ch_in = context.create_channel()
    ch_out = context.create_channel()
    inputs = [Message(i, chunk) for i, chunk in enumerate(chunks)]
    pull_actor, deferred = pull_from_channel(context, ch_out)
    run_actor_network(
        context,
        actors=[
            push_to_channel(context, ch_in, inputs),
            estimator.estimate(context, ch_in, ch_out),
            pull_actor,
        ],
    )
    outputs = deferred.release()
    assert len(outputs) == 1
    return CardinalityEstimate.from_message(outputs[0])


def sample(
    context: Context,
    comm: Communicator,
    chunks: list[TableChunk],
    *,
    max_chunks: int,
    column_indices: tuple[int, ...] = (),
) -> tuple[list[Message], CardinalityEstimate]:
    estimator = CardinalityEstimator(context, comm, tag=0)
    ch_in = context.create_channel()
    ch_sampled = context.create_channel()
    ch_out = context.create_channel()
    inputs = [Message(i, chunk) for i, chunk in enumerate(chunks[:max_chunks])]

    sampled_actor, sampled_deferred = pull_from_channel(context, ch_sampled)
    estimate_actor, estimate_deferred = pull_from_channel(context, ch_out)
    run_actor_network(
        context,
        actors=[
            push_to_channel(context, ch_in, inputs),
            estimator.estimate(
                context,
                ch_in,
                ch_out,
                ch_sampled,
                column_indices,
            ),
            sampled_actor,
            estimate_actor,
        ],
    )
    estimates = estimate_deferred.release()
    assert len(estimates) == 1
    return (
        sampled_deferred.release(),
        CardinalityEstimate.from_message(estimates[0]),
    )


def test_estimate_global_cardinality(
    context: Context, comm: Communicator
) -> None:
    distinct_count = 10_000
    values = np.tile(np.arange(distinct_count, dtype=np.int64), 2)
    result = estimate(
        context,
        comm,
        [make_table(part, context) for part in np.array_split(values, 3)],
    )

    assert result.row_count == values.size * comm.nranks
    assert result.distinct_count == pytest.approx(distinct_count, rel=0.05)


def test_estimate_empty_input(context: Context, comm: Communicator) -> None:
    result = estimate(context, comm, [])

    assert result.row_count == 0
    assert result.distinct_count == 0


def test_estimate_forwards_input(context: Context, comm: Communicator) -> None:
    values = np.arange(100, dtype=np.int64)
    sampled, result = sample(
        context,
        comm,
        [make_table(values, context)],
        max_chunks=2,
    )

    assert len(sampled) == 1
    assert result.row_count == values.size * comm.nranks
    assert result.distinct_count == pytest.approx(values.size, rel=0.05)


def test_precision(context: Context, comm: Communicator) -> None:
    assert CardinalityEstimator(context, comm, tag=7).tag == 7
    assert CardinalityEstimator(context, comm, tag=7).precision == 14
    assert (
        CardinalityEstimator(context, comm, tag=7, precision=12).precision
        == 12
    )


@pytest.mark.parametrize("precision", [3, 19])
def test_invalid_precision(
    context: Context, comm: Communicator, precision: int
) -> None:
    with pytest.raises(
        ValueError, match=r"precision must be in range \[4, 18\]"
    ):
        CardinalityEstimator(context, comm, tag=0, precision=precision)
